### ISSUE TITLE
Little confused with BinActive function

### ISSUE THREAD

**Issue by zzqiuzz** [jiecaoyu/XNOR-Net-PyTorch#23]
For my view with BWN,In the forward pass,we only need to compute alpha and a binary filter B w.r.t its float  one in XnorNet paper,but not to binarize a filter's output which is also the input of next layer.So ,why we need **BinActive** function?And  I think your implementation of MNIST is real a BWN version instead of XNOR version after reading your code.Or maybe i ignore some details . 
Another question is: Datatype to represent 0/1 is still float which can't reduce memory or time in training pass.I'm very confused.
THanks for your help

**Comment by jiecaoyu** [comment]
@zzqiuzz Hi, XNOR-Net has all weights/inputs to be binary. In this case, ```BinActive()``` is used to binarize the input values. The weights are binarized by [```binarizeConvParams()```](https://github.com/jiecaoyu/XNOR-Net-PyTorch/blob/7f84f1bf6921bc99b05bbc869dddf2bd64989a40/MNIST/util.py#L52). My implementation is only for XNOR-Net not BWN.

**Comment by zzqiuzz** [comment]
Oh yes.I delved into your code for more details.I find you just calculate alpha for each conv filter not beta for input tensors.The output of a conv-layer is just fed into a sign() function then followed by a normal conv operation instead of Xnor+bitcounting. Why? 
Am i right? Or please correct my misunderstanding.

And what's this mean below?
https://github.com/jiecaoyu/XNOR-Net-PyTorch/blob/7f84f1bf6921bc99b05bbc869dddf2bd64989a40/MNIST/models/LeNet_5.py#L81

THanks!

**Comment by jiecaoyu** [comment]
All inputs and weights are binarized so the normal conv has exactly same output with xnor+bitcounting.

We will initialize the weights in BN layers to 1.

**Comment by niluanwudidadi** [comment]
@jiecaoyu Hello, I have read your code, it's great.Thank you for your impletion! I have some question.**First,** in binarizeConvParams() function, weights are not binary, there are float value in self.target_modules, I think it's just do the backward in BinConv2d the weights and input are binary, do I have a wrong understanding?
**Second**, I can't understand this code 
```
m = self.target_modules[index].data.norm(1, 3, keepdim=True)\
                    .sum(2, keepdim=True).sum(1, keepdim=True).div(n)
```
Is this is mean of weights?

**Comment by jiecaoyu** [comment]
@niluanwudidadi Hi,

First, in forwarding, weights and inputs to the binary convolution layer are also binary. Binary values do not need to be -1 and +1 exactly. They can also use values of ```[a, -a]``` where ```a``` is a floating-point value.

Second, considering weights of one convolutional layer to have a size of ```Out Channel x In Channel x Kernel Size x Kernel Size```. This function is calculating the mean value of each ```In Channel x Kernel Size x Kernel Size``` part of the weights.

**Comment by niluanwudidadi** [comment]
@jiecaoyu Thank you for your reply! emmm, I have more detail question, in the net, nn.Conv2d is a normal conv layer, so its forward and backward function is float-operation operation, while in BinConv2d, the forward and backward function is reconfiguration, so the input are binarized; and in **`**class BinConv2d(nn.Module):**,`**
 ```
self.conv = nn.Conv2d(input_channels, output_channels,
                kernel_size=kernel_size, stride=stride, padding=padding)
```
is also a simple conv2d function , so the weight in this function is target_modules.data? I'm not familiar to Pytorch, So I'm confused.

More question, I also want to know the function of **meancenterConvParams()**, and the **clampConvParams**() just use for differentiate?

And this 
```
self.target_modules[index].data.norm(1, 3, keepdim=True)\
                    .sum(2, keepdim=True).sum(1, keepdim=True).div(n)
```
is the alpha?  And which is the patitial(sign(W))/ patitial(w) in your code?

**Comment by jiecaoyu** [comment]
Hi @niluanwudidadi ,

1. Yes

2. I am not sure why ```meancenterConvParams()``` is important but it really helps increase the accuracy. ```clampConvParams()``` make sure all weights are limited to 1.0 and, therefore, we will not have "dead" weights which will not be updated anymore.

3. Yes. Multiplication with patitial(sign(W))/ patitial(w) is performed by [line 78 + 79](https://github.com/jiecaoyu/XNOR-Net-PyTorch/blob/7f84f1bf6921bc99b05bbc869dddf2bd64989a40/ImageNet/networks/util.py#L78)

**Comment by niluanwudidadi** [comment]
@jiecaoyu  Thank you very much, I have clear my puzzlement.

**Comment by zzqiuzz** [comment]
Hi! @jiecaoyu I've implemented BWN in caffe. However,  version with `meancenterConvParams()`and `clampConvParams()` is not superior to the one without operations above. Or maybe something wrong happens to my code. I notice that In your code, you perform `clampConvParams()` before weights update right? I think it's more reasonable to clip weights after weights update instead. Thanks!

**Comment by jiecaoyu** [comment]
Hi @zzqiuzz performing ```clampConvParams()``` restrict all weights to be in [-1, 1] and all weights can be updated in the weight update step. Empirically, I don't expect this change to lead to a large accuracy difference. Thanks!
